Enhance OpenID Connect Plugin

dr4hcu5-jan opened the pull request:
This PR enhances the maintainability of the OpenID Connect plugin which is used to authenticate requests to the microservices. Furthermore, the Docker image build has been made more-caching friendly to allow for a better deployment experience

dr4hcu5-jan committed:
trying to debbug api gateway

dr4hcu5-jan committed:
updating provided image tags on gateway

dr4hcu5-jan committed:
updating authorization plugin to stop 500 errors on startup

dr4hcu5-jan committed:
cleaning up code and updating logging levels

dr4hcu5-jan committed:
updating docker builds

dr4hcu5-jan committed:
updating the readme

dr4hcu5-jan committed:
enhancing readme again

dr4hcu5-jan committed:
fixing linebreak error in readme

cptpiepmatz commented on README.md:
```diff
@@ -1,23 +1,28 @@
+to the microservices implemented for the platform.
+Furhtermore, it is extended by a plugin using the 
+[`go-pdk`](https://github.com/Kong/go-pdk) to validate JWTs which are used in a
+standard deplyoment to authenticate and secure requests.
```
```suggestion
standard deployment to authenticate and secure requests.
```

cptpiepmatz commented on README.md:
```diff
@@ -1,23 +1,28 @@
+The API Gateway is autommatically configured using the 
+[gateway-service-watcher](https://github.com/wisdom-oss/gateway-service-watcher)
+which acts as a watchdog to check the deployed containers on the host for their
+association to the WISdoM plaform.
```
```suggestion
association to the WISdoM platform.
```

cptpiepmatz commented on plugins/oidc/access.go:
```diff
@@ -10,118 +10,207 @@ import (
+	// cleanup the discovery endpoint address and remove possible whitespaces
+	discoveryEndpoint = strings.TrimSpace(discoveryEndpoint)
+	if discoveryEndpoint == "" {
+		return "", "", errEmptyEndpoint
+	}
+
+	// now request the openid connect configuration endpoint
+	res, err := http.Get(discoveryEndpoint)
+	if err != nil {
+		return "", "", errors.Join(errDiscoveryRequestFailure, err)
+	}
+
+	// now read the openid connect configuration
+	var openIdConnectConfiguration map[string]interface{}
+	err = json.NewDecoder(res.Body).Decode(&openIdConnectConfiguration)
+	if err != nil {
+		return "", "", errors.Join(err, errDiscoveryUnmarshalFailure)
```
I think I would return `nil` instead of empty strings in the error case, but that is not really an issue.

dr4hcu5-jan replied:
This is the same as the following comment: https://github.com/wisdom-oss/api-gateway/pull/4#discussion_r1470902455

cptpiepmatz commented on plugins/oidc/access.go:
```diff
@@ -10,118 +10,207 @@ import (
+		return "", "", errors.Join(err, errDiscoveryUnmarshalFailure)
+	}
+
+	return openIdConnectConfiguration["userinfo_endpoint"].(string), openIdConnectConfiguration["jwks_uri"].(string), nil
```
The assertion here would crash the application, is that expected behavior?
Maybe do a type check and return an error if that the json has an incorrect format instead.

dr4hcu5-jan replied:
Yes, this type assertion would crash the application. A type check will be added

cptpiepmatz commented on plugins/oidc/access.go:
```diff
@@ -10,118 +10,207 @@ import (
+//
+//	The extracted bearer token, or an empty string if no valid bearer token is found.
+func extractBearerToken(authorizationHeaderValue string) string {
+	regex := regexp.MustCompile(headerRegex)
```
Why did you move the regex compile step into the function, will that be rerun every time?

dr4hcu5-jan replied:
The compile step has been moved down here because sometimes the regex compilation failed. However, the regex compile step can be made to be only executed once, using the following code:
```go
// at the top of the file
const bearerTokenExtractionRegExString = `^(\w+) (\S+)$`
var bearerTokenExtractionRegEx *regexp.Regexp

func extractBearerToken(authorizationHeaderValue string) string {
    if bearerTokenExtractionRegEx == nil {
      bearerTokenExtractionRegEx = regexp.MustCompile(bearerTokenExtractionRegExString)
    }
    // rest of the function...
}
```
This way the regex will only compiled once, but only if the function is called

cptpiepmatz replied:
This is a good alternative. But why would a hard-coded regex fail to compile sometimes?

dr4hcu5-jan replied:
Local testing showed that not the regex failed to compile, but trying to use the regex resulted in a panic due to a nil pointer dereference. This happened if there was a request in the backlog of the gateway and the plugin is not started fully which happens if the configuration of the plugin has been changed. This is an edge case and was only reproduced by loading the gateway with about 25000 requests in its backlog and changing the gateway every 2.5 seconds.

cptpiepmatz replied:
Huh, then we could just move the regex compile up again or not?

dr4hcu5-jan replied:
No, because using the compile step in this way ensures that the regex is not `nil` before usage and if the regex is `nil` it compiles the regex

dr4hcu5-jan replied:
Furthermore, the regex is now compiled only once and not at every access which reduces the time for handling the request

cptpiepmatz commented on plugins/oidc/access.go:
```diff
@@ -10,118 +10,207 @@ import (
+func extractBearerToken(authorizationHeaderValue string) string {
+	regex := regexp.MustCompile(headerRegex)
+	matches := regex.FindStringSubmatch(authorizationHeaderValue)
+	if len(matches) == 3 {
+		return matches[2]
+	}
```
This works but a [named group](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Named_capturing_group) could be clearer to use.

dr4hcu5-jan replied:
Using a named group will only make the regex more illegible since there is no function in the standard package for regular expressions that handles returning named groups directly. Therefore, in my opinion using a named group will not increase the clarity of this part
> https://pkg.go.dev/regexp#pkg-index

cptpiepmatz commented on plugins/oidc/access.go:
```diff
@@ -10,118 +10,207 @@ import (
+	if len(matches) == 3 {
+		return matches[2]
+	}
+	return ""
```
Why return an empty string instead of `nil`. I think a `nil` would be clearer. (Or some optional type.)

dr4hcu5-jan replied:
Returning `nil` would require using `*string` as return type and requires a more thorough check of the return type in the code later on, since a `*string` cannot be used directly in a `if` statement to check the length and would require the following syntax making the function using the return value longer:
```go
if accessToken == nil {
  // raise error here
}

if strings.Trim(*accessToken) == "" {
  // raise another error here
}
```

cptpiepmatz commented on plugins/oidc/access.go:
```diff
@@ -10,118 +10,207 @@ import (
+	return ""
+}
+
+// Access is executed every time the kong gateway receives a request. Since ther
```
```suggestion
// Access is executed every time the kong gateway receives a request. Since the
```

cptpiepmatz commented on plugins/oidc/access.go:
```diff
@@ -10,118 +10,207 @@ import (
-		// since the matches array needs to contain three entries, the value is
-		// invalid and therefore the user is not authorized
+
+	userinfoEndpoint, jwksEndpoint, err := discoverEndpoints(c.DiscoveryUri)
```
The endpoint discovery always runs a GET request which takes some time. Maybe this is a bit too rigorous.

dr4hcu5-jan replied:
True, however this is the only way a configuration change during the plugin runtime can be caught since the plugin does not need to be restarted if the plugin configuration has been changed and caching these values may result in an validation error further down in the code

cptpiepmatz commented on plugins/oidc/access.go:
```diff
@@ -10,118 +10,207 @@ import (
 		}
-		headers := map[string][]string{
-			"WWW-Authenticate": {`Bearer scope="openid profile email", error="invalid_request", error_description="Authorization header malformed'"`},
+		switch {
+		case errors.Is(err, errEmptyEndpoint):
+			response = GatewayError{
+				ErrorCode:        "gateway.OPENID_CONNECT_INVALID_DISCOVERY_ENDPOINT",
+				ErrorTitle:       "Invalid Discovery Endpoint",
+				ErrorDescription: "The configured OpenID Connect discovery endpoint is invalid: " + err.Error(),
+			}
+		case errors.Is(err, errDiscoveryRequestFailure):
+			response = GatewayError{
+				ErrorCode:        "gateway.OPENID_CONNECT_DISCOVERY_REQUEST_FAILURE",
+				ErrorTitle:       "Discovery Request Failure",
+				ErrorDescription: "The discovery request needed for access token validation failed: " + err.Error(),
+			}
+		case errors.Is(err, errDiscoveryUnmarshalFailure):
+			response = GatewayError{
+				ErrorCode:        "gateway.OPENID_CONNECT_DISCOVERY_RESPONSE_PARSE_ERROR",
+				ErrorTitle:       "Discovery Response Parsing Error",
+				ErrorDescription: "The discovery response sent by the configured endpoint could not be parsed: " + err.Error(),
+			}
```
This is nice :+1:

cptpiepmatz commented on plugins/oidc/access.go:
```diff
@@ -10,118 +10,207 @@ import (
+
+// Access is executed every time the kong gateway receives a request. Since ther
 // plugin may be restarted at any moment
 func (c *Configuration) Access(kong *pdk.PDK) {
```
This entire function is very long. You have multiple phases/steps. Maybe reduce this function to calls into subfunctions that each handle the steps. 

Maybe you can also shorten the error handling by constructing the error type in the subfunctions, by this you can something more like this:
```go
validated, err := validateSomething()
if err != nil {
  return nil, err
}
```

dr4hcu5-jan committed:
Update README.md

Co-authored-by: Piepmatz <3590829+cptpiepmatz@users.noreply.github.com>

dr4hcu5-jan committed:
Update README.md

Co-authored-by: Piepmatz <3590829+cptpiepmatz@users.noreply.github.com>

dr4hcu5-jan committed:
adding .gitignore

dr4hcu5-jan committed:
changing regex and compilation of it

the regex now is marked as case insensitive and will only extract a bearer token and not other types of tokens

dr4hcu5-jan committed:
excluding .exe files

dr4hcu5-jan committed:
implementing type checking on endpoint discovery

dr4hcu5-jan committed:
Merge branch 'dev' of https://github.com/wisdom-oss/api-gateway into dev

dr4hcu5-jan committed:
updating readme to point to watchdog for auth config

cptpiepmatz commented on plugins/oidc/README.md:
```diff
@@ -20,5 +20,7 @@ These responses can optionally be cached using a redis database to further
+The configuration is done by the 
+[watchdog](https://github.com/wisdom-oss/watchdog) on a service-by-service
+basis.
+Please refer to the documentation of the watchdog for further information
```
```suggestion
Please refer to the documentation of the watchdog for further information.
```

dr4hcu5-jan committed:
Update plugins/oidc/README.md

Co-authored-by: Piepmatz <3590829+cptpiepmatz@users.noreply.github.com>
